crates/pallet-domains/src/bundle_storage_fund.rs

```diff
@@ -0,0 +1,107 @@
+//! Bundle storage fund
+
+use crate::{BalanceOf, Config, Operators};
+use codec::{Decode, Encode};
+use frame_support::traits::fungible::Mutate;
+use frame_support::traits::tokens::{Fortitude, Precision};
+use frame_support::traits::Get;
+use frame_support::PalletError;
+use scale_info::TypeInfo;
+use sp_domains::OperatorId;
+use sp_runtime::traits::{AccountIdConversion, CheckedSub, Zero};
+use sp_runtime::Perbill;
+use sp_std::collections::btree_map::BTreeMap;
+use subspace_runtime_primitives::StorageFeeInterface;
+
+/// Bundle storage fund specific errors
+#[derive(TypeInfo, Encode, Decode, PalletError, Debug, PartialEq)]
+pub enum Error {
+    FailedToDeriveStorageFundAccount,
+    BundleStorageFeePayment,
+    BalanceUnderflow,
+    MintBalance,
+}
+
+/// The type of system account being created.
+#[derive(Encode, Decode)]
+pub enum AccountType {
+    StorageFund,
+}
+
+/// Return the bundle storage fund account of the given operator.
+pub fn storage_fund_account<T: Config>(id: OperatorId) -> Result<T::AccountId, Error> {
+    T::PalletId::get()
+        .try_into_sub_account((AccountType::StorageFund, id))
+        .ok_or(Error::FailedToDeriveStorageFundAccount)
+}
+
+/// Charge the bundle storage fee from the operator's bundle storage fund
+pub fn charge_bundle_storage_fee<T: Config>(
+    operator_id: OperatorId,
+    bundle_size: u32,
+) -> Result<(), Error> {
+    if bundle_size.is_zero() {
+        return Ok(());
+    }
+
+    let storage_fund_acc = storage_fund_account::<T>(operator_id)?;
+    let storage_fee = T::StorageFeeInterface::transaction_byte_fee() * bundle_size.into();
+
+    T::Currency::burn_from(
```

Comment:
@dariolina do we need to burn the storage fees here and then mint back again ? Can we instead do a transfer of funds ?
Burning will cause the total issuance to always fluctuate

Reply:
This is exactly how every other transaction pays the fee where the tx fee is burnt and noted in `pallet-transaction-fees` then the total tx fee of the current block will be mint into the block author's reward address at block finalization.

Reply:
Wouldn't it be better to instead use an protocol controlled account to transfer those fees temporarily and then transfer back to block author instead of burning and there by fluctuating the total supply of SSC in every block. This is out of scope for this PR but I think that would be a better way to handle this on consensus chain. Thoughts @nazar-pc ?